File: Script/Python/src/STG/Stage/Test/Stage1/Enemy/Idiot01.py
'''
Created on 2010/11/23

@author: t
'''

import Std
import STG.Ctrl as Ctrl
import STG.Auxs as Auxs
from STG.Enemy.BaseEnemy import Base

class Idiot01(Base):
	SIZE = Std.Vector2DF(48, 48)
	EXPLOSION_SIZE = Std.Vector2DF(96, 96)
	MAX_HP = 10
	
	def __init__(self, pos, res, scoreManager):
		super().__init__(
			Std.Hit.RectI(pos, self.SIZE, True), self.MAX_HP)
		self.scoreManager = scoreManager
		
		self.locator = Std.Locator.LinearF(pos)
		self.locator.speed.setUnitVector(90)
		self.locator.speed *= 4
		
		self.drawParam = Std.Sprite.DrawParam()
		self.drawParam.texture = res.icon2
		self.drawParam.src = Std.RectF(0, 0, 256, 256)
		self.drawParam.dst = pos.makeRect(self.SIZE, True)
		self.drawParam.priority = Ctrl.DrawPriority.enemy
		
		self.launchInterval = Std.RepeatCounter(16)
	
	@property
	def position(self):
		return self.locator.position
	def applyExternalForce(self, vec):
		return self.locator.applyExternalForce(vec)
	
	def onUpdate(self):
		self.locator.update()
		self.drawParam.dst = self.locator.position.makeRect(
			self.SIZE, True)
		
		self.launchInterval.update()
		for _ in range(self.launchInterval.repeatCount):
			self.launchSimpleBullet()
	def onDraw(self):
		Std.Sprite.draw(self.drawParam)
	
	def onErase(self):
		self.onDestructed()
	def onDestructed(self):
		Auxs.createExplosion(
			self.position, self.EXPLOSION_SIZE)
		self.scoreManager.addScore(self.maxHp)
	def onDamaged(self, damage):
		self.scoreManager.addRate(damage * 0.01)
	def onDestructedByDamage(self, damage):
		self.onDestructed()
		
		item = Auxs.Item.PowerUp(self.locator.position)
		def effectCallback(result):
			if not result:
				self.scoreManager.addScore(1000)
		item.setEffectCallback(effectCallback)
		Ctrl.Actor.registerItem(item)
	
	def launchSimpleBullet(self):
		pos = self.position
		angle = 90 - 20
		
		for i in range(3):
			locator = Std.Locator.LinearF(pos)
			locator.speed.setUnitVector(angle)
			locator.speed *= 6
			
			hitRadius = 3
			
			drawParam = Auxs.createBulletRedDrawParam(
				pos, Std.Vector2DF(14, 14))
			
			bullet = Auxs.Bullet.Linear(locator, hitRadius, drawParam)
			Ctrl.Actor.registerBullet(bullet)
			
			angle = 90 + 20*i

File: STGProject/Source/Mdl/STG/Stage/Test/TestStage1/Enemy/Idiot01.cpp
#include "PCH.h"
#include "Idiot01.h"

#include "Ctrl/STG/Actors.h"
#include "Mdl/STG/Stage/Test/Resource.h"
#include "Mdl/STG/Stage/Test/StageSetCore.h"
#include "View/STG/DrawPriority.h"

using namespace Game;
using namespace Mdl::STG;
using namespace Stage::Test::TestStage1::Enemy;
using namespace Util::STG;
using namespace Selene;

namespace
{
	static const Vector2DI EN_SIZE( 48, 48 );
	static const Vector2DF EXPLOSION_SIZE( 96, 96 );
	static const float HP = 5;
	static const unsigned WAIT_FRAME_NUM = 10;
	static const Hit::RectI LAUNCH_AREA( 0, 0, 440, 360, false );
}


Idiot01::Idiot01( PStageSet pStageSet, const Util::STG::Vector2DF &initPos, 
	const Util::STG::Vector2DF &targetPos, float speed )
	: Base( pStageSet->GetMode(), Hit::RectI( initPos, EN_SIZE ), HP )
	, mpStageSet( pStageSet )
	, mLocator( initPos )
	, mInitPos( initPos )
	, mTargetPos( targetPos )
	, mArrivalTargetPos( false )
	, mWaitCount( 0 )
	, mSpeed( speed )
	, mAuxs( pStageSet->GetMode() )
{
	mLocator.GetSpeed().SetUnitVector( 
		mLocator.GetPosition().GetAngle( mTargetPos ) ) *= mSpeed;

	mDrawParam.SetTexture( mpStageSet->GetResource()->GetIcon2() );
	mDrawParam.SetSrc( Resource::ICON2.GetRect() );
	mDrawParam.SetDst( mLocator.GetPosition().MakeRect( EN_SIZE ) );
	mDrawParam.SetPriority( View::STG::PRI_ENEMY );
}


Vector2DF Idiot01::GetPosition() const
{
	return mLocator.GetPosition();
}

Vector2DF Idiot01::ApplyExternalForce( const Vector2DF &vec )
{
	return mLocator.ApplyExternalForce( vec );
}


void Idiot01::OnUpdate()
{
	mLocator.Update();

	if( !mArrivalTargetPos )
	{
		if( mLocator.GetPosition().GetDistance( mTargetPos ) <= 
			mLocator.GetSpeed().GetAbs() )
		{
			mLocator.GetAccel() = 0;
			mLocator.GetSpeed() = 0;
			mLocator.GetPosition() = mTargetPos;

			mArrivalTargetPos = true;
		}
		else if( mLocator.GetPosition().GetDistance( mTargetPos ) <= 50.0f )
		{
			if( mLocator.GetSpeed().GetAbs() > 1.0f )
			{
				mLocator.GetAccel().SetUnitVector( 
					mLocator.GetSpeed().GetAngle() + 180.0f ) *= mSpeed / 20.0f;
			}
			else
			{
				mLocator.GetAccel() = 0;
				mLocator.GetSpeed().Normalize();
			}
		}
	}
	else
	{
		if( mWaitCount == WAIT_FRAME_NUM )
		{
			if( mLocator.GetSpeed().GetAbs() < mSpeed )
			{
				mLocator.GetAccel().SetUnitVector( 
					mInitPos.GetAngle( mTargetPos ) ) *= mSpeed / 50.0f;
			}
			else
			{
				mLocator.GetAccel() = 0;
				mLocator.GetSpeed().SetUnitVector( 
					mInitPos.GetAngle( mTargetPos ) ) *= mSpeed;
			}
		}
		else
		{
			mWaitCount++;
		}
	}

	mLaunchItv.UpdateState();
	if( Base::IsHit( LAUNCH_AREA ) )
	{
		for( unsigned i = 0; i < mLaunchItv.GetRepeatCount(); i++ )
		{
			Launch();
		}
	}

	mDrawParam.SetDst( mLocator.GetPosition().MakeRect( EN_SIZE ) );
}

void Idiot01::OnDraw() const
{
	Util::Sprite::Manager::Draw( mDrawParam );
}

void Idiot01::OnErase()
{
	OnDestruct();
}

void Idiot01::OnDestruct()
{
	mpStageSet->AddScore( static_cast<Uint64>( mMaxHP ) );

	mAuxs.CreateExplosion( mLocator.GetPosition(), EXPLOSION_SIZE );
}

void Idiot01::OnDamaged( float damage )
{
	mpStageSet->AddRate( damage * 0.01f );
}

void Idiot01::OnDestruct( float hp )
{
	OnDestruct();
}